iothub_client/tests/common_dt_e2e/iothubclient_common_dt_e2e.c

```diff
@@ -10,568 +10,728 @@
+        received_twin_data->cb_payload_size = size;
+        received_twin_data->received_callback = true;
+        Unlock(received_twin_data->lock);
     }
 }
 
-
-
-
-static const char *COMPLETE_DESIRED_PAYLOAD_FORMAT = "{\"properties\":{\"desired\":{\"integer_property\": %d, \"string_property\": \"%s\", \"array\": [%d, \"%s\"]}}}";
-static char *malloc_and_fill_desired_payload(const char *string, int aint)
+static void _set_twin_callback(IOTHUB_CLIENT_DEVICE_TWIN_CALLBACK twin_callback,
+                               RECEIVED_TWIN_DATA* twin_data_context)
 {
-    size_t  length = snprintf(NULL, 0, COMPLETE_DESIRED_PAYLOAD_FORMAT, aint, string, aint, string);
-    char   *retValue = (char *) malloc(length + 1);
-    if (retValue == NULL)
+    IOTHUB_CLIENT_RESULT result;
+
+    if (iothub_moduleclient_handle)
     {
-        LogError("malloc failed");
+        result = IoTHubModuleClient_SetModuleTwinCallback(iothub_moduleclient_handle, twin_callback,
+                                                          twin_data_context);
     }
     else
     {
-        (void) sprintf(retValue, COMPLETE_DESIRED_PAYLOAD_FORMAT, aint, string, aint, string);
+        result = IoTHubDeviceClient_SetDeviceTwinCallback(iothub_deviceclient_handle, twin_callback,
+                                                          twin_data_context);
     }
-    return retValue;
+
+    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result,
+                     "IoTHub(Device|Module)Client_SetDeviceTwinCallback failed.");
 }
 
-static char * malloc_and_copy_unsigned_char(const unsigned char* payload, size_t size)
+static void _get_twin_async(IOTHUB_CLIENT_DEVICE_TWIN_CALLBACK twin_callback,
+                            RECEIVED_TWIN_DATA* received_twin_data)
 {
-    char *retValue;
-    if (payload == NULL)
+    IOTHUB_CLIENT_RESULT result;
+
+    if (iothub_moduleclient_handle)
     {
-        LogError("invalid parameter payload");
-        retValue = NULL;
+        result = IoTHubModuleClient_GetTwinAsync(iothub_moduleclient_handle, twin_callback,
+                                                 received_twin_data);
     }
-    else if (size < 1)
+    else
     {
-        LogError("invalid parameter size");
-        retValue = NULL;
+        result = IoTHubDeviceClient_GetTwinAsync(iothub_deviceclient_handle, twin_callback,
+                                                 received_twin_data);
     }
-    else
+
+    ASSERT_ARE_EQUAL(IOTHUB_CLIENT_RESULT, IOTHUB_CLIENT_OK, result,
+                     "IoTHub(Device|Module)Client_GetTwinAsync failed.");
+}
+
+static void _receive_twin_loop(RECEIVED_TWIN_DATA* received_twin_data, DEVICE_TWIN_UPDATE_STATE expected_update_state)
+{
+    time_t begin_operation = time(NULL);
+    time_t now_time;
+
+    LogInfo("_receive_twin_loop(): Entering loop.\n");
+
+    while (now_time = time(NULL), difftime(now_time, begin_operation) < MAX_CLOUD_TRAVEL_TIME)
```

Comment:
Nit - could we put a parattheses around the difftime(...) <MAX....  I get this should be OK but whenever I see commas I start thinking about order of operation stuff and the paren would make it clearer.  Similar feedback for other place that does this.